cloud/services/virtualmachines/virtualmachines.go

```diff
@@ -158,6 +159,21 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 		virtualMachine.Identity = &compute.VirtualMachineIdentity{
 			Type: compute.ResourceIdentityTypeSystemAssigned,
 		}
+	} else if vmSpec.Identity == infrav1.VMIdentityUserAssigned {
+		if len(vmSpec.UserAssignedIdentities) == 0 {
+			return errors.Wrapf(err, "cannot create VM: The user-assigned identity provider ids must not be null or empty for 'UserAssigned' identity type.")
```

Comment:
I would adjust the error text similar to my suggestion above. Users probably aren't interested in null vs empty, so I'd stick with stronger language about them being required.

Reply:
What happens when you return an error from here? If it bubbles all the way up to the main Reconcile method and is returned from there, it will be logged in the pod's log and not visible to the end user.

Reply:
yes it will be logged in the pod's log, but I also added the same validation in the webhook, but thought I should leave it here anyways, would it be better to remove it from here?